File: forwarder/app/src/main/java/com/codey/forwarder/ForwarderNotifier.java
package com.codey.forwarder;

import android.app.Notification;
import android.os.Bundle;
import android.service.notification.StatusBarNotification;

import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

final class ForwarderNotifier {
    private static final String WHATSAPP_PACKAGE = "com.whatsapp";
    private static final String WHATSAPP_BUSINESS_PACKAGE = "com.whatsapp.w4b";

    private ForwarderNotifier() {}

    static boolean shouldForwardPackage(String packageName, boolean forwardBusiness) {
        return WHATSAPP_PACKAGE.equals(packageName) ||
            (forwardBusiness && WHATSAPP_BUSINESS_PACKAGE.equals(packageName));
    }

    static ForwardedNotification buildForwardedNotification(StatusBarNotification sbn) {
        Bundle extras = sbn.getNotification().extras;
        String title = readCharSequence(extras, Notification.EXTRA_TITLE);
        String text = readCharSequence(extras, Notification.EXTRA_TEXT);
        String bigText = readCharSequence(extras, Notification.EXTRA_BIG_TEXT);
        String subText = readCharSequence(extras, Notification.EXTRA_SUB_TEXT);
        String conversationTitle = readCharSequence(extras, Notification.EXTRA_CONVERSATION_TITLE);
        String body = bigText != null ? bigText : text;
        long receivedAt = sbn.getPostTime() > 0L ? sbn.getPostTime() : System.currentTimeMillis();
        String key = sbn.getKey();
        String notificationId = key != null
            ? key
            : sbn.getPackageName() + ":" + sbn.getId() + ":" + receivedAt;

        return new ForwardedNotification(
            sbn.getPackageName(),
            notificationId,
            title,
            body,
            text,
            bigText,
            subText,
            conversationTitle,
            receivedAt
        );
    }

    static JSONObject buildForwarderPayload(
        ForwarderSettings settings,
        ForwardedNotification notification
    ) throws Exception {
        JSONObject rawPayload = new JSONObject()
            .put("source", "codey-forwarder")
            .put("packageName", notification.packageName)
            .put("notificationId", notification.notificationId)
            .put("postTime", notification.receivedAt);
        ForwarderConfig.putOptional(rawPayload, "android.text", notification.text);
        ForwarderConfig.putOptional(rawPayload, "android.bigText", notification.bigText);
        ForwarderConfig.putOptional(rawPayload, "android.subText", notification.subText);
        ForwarderConfig.putOptional(
            rawPayload,
            "android.conversationTitle",
            notification.conversationTitle
        );

        JSONObject payload = new JSONObject()
            .put("source", "codey-forwarder")
            .put("deviceId", settings.deviceId)
            .put("packageName", notification.packageName)
            .put("notificationId", notification.notificationId)
            .put("receivedAt", notification.receivedAt)
            .put("rawPayload", rawPayload);
        ForwarderConfig.putOptional(payload, "sender", notification.title);
        ForwarderConfig.putOptional(
            payload,
            "chatName",
            notification.conversationTitle != null
                ? notification.conversationTitle
                : notification.title
        );
        ForwarderConfig.putOptional(payload, "title", notification.title);
        ForwarderConfig.putOptional(payload, "body", notification.body);
        return payload;
    }

    static HttpResult postNotificationPayload(String webhookUrl, JSONObject payload)
        throws Exception {
        HttpURLConnection connection = (HttpURLConnection) new URL(webhookUrl).openConnection();
        connection.setRequestMethod("POST");
        connection.setConnectTimeout(10_000);
        connection.setReadTimeout(10_000);
        connection.setDoOutput(true);
        connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
        connection.setRequestProperty("Accept", "application/json");

        try (OutputStreamWriter writer = new OutputStreamWriter(
            connection.getOutputStream(),
            StandardCharsets.UTF_8
        )) {
            writer.write(payload.toString());
        }

        int statusCode = connection.getResponseCode();
        InputStream stream = statusCode >= 200 && statusCode <= 299
            ? connection.getInputStream()
            : connection.getErrorStream();
        String responseBody = readStream(stream);
        connection.disconnect();
        return new HttpResult(statusCode, responseBody);
    }

    private static String readStream(InputStream stream) throws Exception {
        if (stream == null) {
            return "";
        }

        try (InputStream input = stream; ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[4096];
            int read;
            while ((read = input.read(buffer)) != -1) {
                output.write(buffer, 0, read);
            }
            return output.toString(StandardCharsets.UTF_8.name());
        }
    }

    private static String readCharSequence(Bundle extras, String key) {
        CharSequence value = extras.getCharSequence(key);
        return value == null ? null : value.toString();
    }

    static final class HttpResult {
        final int statusCode;
        final String responseBody;

        HttpResult(int statusCode, String responseBody) {
            this.statusCode = statusCode;
            this.responseBody = responseBody;
        }
    }
}

File: forwarder/app/src/main/java/com/codey/forwarder/ForwarderNotifier.kt
package com.codey.app

import android.app.Notification
import android.os.Bundle
import android.service.notification.StatusBarNotification
import org.json.JSONObject
import java.io.InputStream
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

object ForwarderNotifier {
    private const val WHATSAPP_PACKAGE = "com.whatsapp"
    private const val WHATSAPP_BUSINESS_PACKAGE = "com.whatsapp.w4b"

    fun shouldForwardPackage(packageName: String?, forwardBusiness: Boolean): Boolean {
        return WHATSAPP_PACKAGE == packageName ||
            (forwardBusiness && WHATSAPP_BUSINESS_PACKAGE == packageName)
    }

    fun buildForwardedNotification(sbn: StatusBarNotification): ForwardedNotification {
        val extras = sbn.notification.extras
        val title = readCharSequence(extras, Notification.EXTRA_TITLE)
        val text = readCharSequence(extras, Notification.EXTRA_TEXT)
        val bigText = readCharSequence(extras, Notification.EXTRA_BIG_TEXT)
        val subText = readCharSequence(extras, Notification.EXTRA_SUB_TEXT)
        val conversationTitle = readCharSequence(extras, Notification.EXTRA_CONVERSATION_TITLE)
        val body = bigText ?: text
        val receivedAt = if (sbn.postTime > 0L) sbn.postTime else System.currentTimeMillis()
        val notificationId = sbn.key ?: "${sbn.packageName}:${sbn.id}:$receivedAt"

        return ForwardedNotification(
            packageName = sbn.packageName,
            notificationId = notificationId,
            title = title,
            body = body,
            text = text,
            bigText = bigText,
            subText = subText,
            conversationTitle = conversationTitle,
            receivedAt = receivedAt
        )
    }

    fun buildForwarderPayload(
        settings: ForwarderSettings,
        notification: ForwardedNotification
    ): JSONObject {
        val rawPayload = JSONObject()
            .put("source", "codey-app")
            .put("packageName", notification.packageName)
            .put("notificationId", notification.notificationId)
            .put("postTime", notification.receivedAt)
        ForwarderConfig.putOptional(rawPayload, "android.text", notification.text)
        ForwarderConfig.putOptional(rawPayload, "android.bigText", notification.bigText)
        ForwarderConfig.putOptional(rawPayload, "android.subText", notification.subText)
        ForwarderConfig.putOptional(
            rawPayload,
            "android.conversationTitle",
            notification.conversationTitle
        )

        val payload = JSONObject()
            .put("source", "codey-app")
            .put("deviceId", settings.deviceId)
            .put("packageName", notification.packageName)
            .put("notificationId", notification.notificationId)
            .put("receivedAt", notification.receivedAt)
            .put("rawPayload", rawPayload)
        ForwarderConfig.putOptional(payload, "sender", notification.title)
        ForwarderConfig.putOptional(
            payload,
            "chatName",
            notification.conversationTitle ?: notification.title
        )
        ForwarderConfig.putOptional(payload, "title", notification.title)
        ForwarderConfig.putOptional(payload, "body", notification.body)
        return payload
    }

    fun postNotificationPayload(
        webhookUrl: String,
        payload: JSONObject,
        bearerToken: String? = null
    ): HttpResult {
        val connection = URL(webhookUrl).openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
        connection.setRequestProperty("Accept", "application/json")
        if (!bearerToken.isNullOrBlank()) {
            connection.setRequestProperty("Authorization", "Bearer ${bearerToken.trim()}")
        }

        OutputStreamWriter(connection.outputStream, StandardCharsets.UTF_8).use { writer ->
            writer.write(payload.toString())
        }

        val statusCode = connection.responseCode
        val stream = if (statusCode in 200..299) {
            connection.inputStream
        } else {
            connection.errorStream
        }
        val responseBody = readStream(stream)
        connection.disconnect()
        return HttpResult(statusCode, responseBody)
    }

    fun buildCodeyWebIngestUrl(settings: ForwarderSettings): String {
        var baseUrl = settings.codeyBaseUrl.trim()
        if (baseUrl.isEmpty()) {
            return settings.webhookUrl
        }
        while (baseUrl.endsWith("/")) {
            baseUrl = baseUrl.dropLast(1)
        }
        return "$baseUrl/api/ingest/whatsapp-notification"
    }

    fun resolveNotificationWebhookUrl(settings: ForwarderSettings): String {
        return if (settings.deviceToken.trim().isEmpty()) {
            settings.webhookUrl
        } else {
            buildCodeyWebIngestUrl(settings)
        }
    }

    private fun readStream(stream: InputStream?): String {
        return stream?.use { input ->
            input.readBytes().toString(StandardCharsets.UTF_8)
        }.orEmpty()
    }

    private fun readCharSequence(extras: Bundle, key: String): String? {
        return extras.getCharSequence(key)?.toString()
    }

    data class HttpResult(
        val statusCode: Int,
        val responseBody: String
    )
}
